### ISSUE TITLE
Freeing netmap buffers in a different order than getting them

### ISSUE THREAD

**Issue by jmtilli** [luigirizzo/netmap#475]
Netmap has rings of e.g. 1024 slots, which contain pointers to buffers. It is usually expected that the application reads them by moving the `cur` pointer forwards up to the `tail` pointer, and deallocates them by moving the `head` pointer forwards.

Now, my question is related to zero-copy input (NB: just zero-copy input, not zero-copy output, not full zero-copy!) applications that need to do QoS, meaning packets may not be transmitted in exactly the same order they were received. This means buffers may not be freed in exactly the same order they were allocated.

Is it possible to free buffers in a different order by manipulating the `buf_idx` pointers of slots that are sent back to the kernel for reuse? I assume if this is at all possible, you probably have to set the `NS_BUF_CHANGED` flag.

Of course, if a ring has only 1024 slots, the application can never hold more than 1024 packets in the memory at a time. Probably the wisest thing to do would be to limit the number of packets to much lower than 1024, e.g. 512.

If this is at all possible, does the `NS_BUF_CHANGED` reduce performance so much that it doesn't make sense to do this?

If this is not supported, it would be a very good idea to support this use case. Many professional applications need to do QoS and thus to transmit (and free) packets in a different order than they were received in.

**Comment by giuseppelettieri** [comment]
Yes, the general idea is that you receive buffers when ```tail``` moves and you return buffers by moving ```head```. There is no obligation to return the same buffers that you received, and indeed this is how zero-copy transfer is implemented in, e.g., ```lb``` and ```bridge```. You do need to set ```NS_BUF_CHANGED``` when you manipulate ```buf_idx```. Beware that also the kernel may set this flag for the buffers it gives to you, so there is no guarantee that you always receive the same buffers in each run through the ring.

You are also not limited to the buffers that are contained in the ring, since you can ask for extra buffers when you register a port. Take a look at ```lb``` for an example of this.

As far as performance is concerned, swapping buffers may have adverse effects on cache and TLB misses, and it may cause the kernel to perform some more work for some combinations of netmap port  type and OS. Most of the time this cost is negligible, but you have to try.
